core/includes/install.core.inc

```diff
@@ -1454,6 +1451,351 @@ function install_select_language_form($form, &$form_state, $files) {
+ *   original name. If the path contains a filename as well, that one will be
+ *   used instead.
+ *
+ * @return bool
+ *   TRUE on success, FALSE on failure.
+ */
+function install_retrieve_file($uri, $destination) {
+  $parsed_url = parse_url($uri);
+
+  if (is_dir(backdrop_realpath($destination))) {
+    // Prevent URIs with triple slashes when gluing parts together.
+    $path = str_replace('///', '//', "$destination/") . backdrop_basename($parsed_url['path']);
+  }
+  else {
+    $path = $destination;
+  }
+
+  try {
+    $response = backdrop_http_request($uri);
+    if ($response->code != 200) {
+      return FALSE;
+    }
+    $data = $response->data;
+    if (empty($data)) {
+      return FALSE;
+    }
+  }
+  catch (TransferException $e) {
+    return FALSE;
+  }
+  return file_put_contents($path, $data) !== FALSE;
+}
+
+/**
+ * Checks if the localization server can be contacted.
+ *
+ * @param string $uri
+ *   The URI to contact.
+ *
+ * @return string
+ *   TRUE if the URI was contacted successfully, FALSE if not.
+ */
+function install_check_localization_server($uri) {
+  $response = backdrop_http_request($uri);
+  if ($response->code != 200) {
+    return FALSE;
+  }
+  return TRUE;
+}
+
+/**
+ * Checks installation requirements and reports any errors.
+ *
+ * @param string $langcode
+ *   Language code to check for download.
+ * @param string $server_pattern
+ *   Server access pattern (to replace language code, version number, etc. in).
+ *
+ * @return array
+ *   Requirements compliance array. If the translation cannot be downloaded this
+ *   will contain a requirements error with detailed information.
+ */
+function install_check_translations($langcode, $server_pattern) {
+  $requirements = [];
+
+  $readable = FALSE;
+  $writable = FALSE;
+  // @todo: Make this configurable.
+  $files_directory = BACKDROP_ROOT . '/files';
+  $translations_directory = BACKDROP_ROOT . '/files/translations';
+  $translations_directory_exists = FALSE;
+  $online = FALSE;
+
+  // First attempt to create or make writable the files directory.
+  file_prepare_directory($files_directory, FILE_CREATE_DIRECTORY | FILE_MODIFY_PERMISSIONS);
+
+  // Then, attempt to create or make writable the translations directory.
+  file_prepare_directory($translations_directory, FILE_CREATE_DIRECTORY | FILE_MODIFY_PERMISSIONS);
+
+  // Get values so the requirements errors can be specific.
+  if (backdrop_verify_install_file($translations_directory, FILE_EXIST, 'dir')) {
+    $readable = is_readable($translations_directory);
+    $writable = is_writable($translations_directory);
+    $translations_directory_exists = TRUE;
+  }
+
+  $version = BACKDROP_VERSION;
+
+  // For dev releases, remove the '-dev' part and trust the translation server
+  // to fall back to the latest stable release for that branch.
+  // @see locale_translation_build_projects()
+  if (preg_match("/^(\d+\.).*-dev$/", $version, $matches)) {
```

Comment:
Our pre-releases are currently named like `1.x.y-preview`, so lets change this to account for that:
```suggestion
  if (preg_match("/^(\d+\.).*-(dev|preview)$/", $version, $matches)) {
```

...either that ^^, or this if we don't want to support `-dev` as a version suffix:
```suggestion
  if (preg_match("/^(\d+\.).*-preview$/", $version, $matches)) {
```

Reply:
...actually scratch the previous comment. It is valid in theory, but our localization server generates .po files for minor versions only as it is currently - not for any `.x` version. So we have files like:
- `https://localize.backdropcms.org/files/l10n_packager/all/backdropcms/backdropcms-1.26.ar.po`
- `https://localize.backdropcms.org/files/l10n_packager/all/backdropcms/backdropcms-1.27.ar.po`
- `https://localize.backdropcms.org/files/l10n_packager/all/backdropcms/backdropcms-1.28.ar.po`

So, we only need to grab the first two digitas and dots. Something like this should work:
```suggestion
  if (preg_match("/^(\d+\.\d+\.).*$/", $version, $matches)) {
```

Reply:
Your first suggestion here is valid: we have `-dev` versions prior to release, as well as `-preview` which I didnt know about. And localize does produce `-1.x` files.